src/pcap/bosh.go

```diff
@@ -1,55 +1,414 @@
+	})
+
+	if err != nil || !token.Valid {
+		return false, err
+	}
+
+	if claims, claimsOk := token.Claims.(jwt.MapClaims); claimsOk {
+		if scopes, ok := claims["scope"].([]interface{}); ok {
+			for _, scope := range scopes {
+				if scope.(string) == br.config.TokenScope {
+					return true, nil
+				}
+			}
+		}
+	}
+
+	return false, fmt.Errorf("could not find scope %q in token claims", br.config.TokenScope)
+}
+
+// parseRsaToken uses the token information for RSA signed JWT tokens and retrieves
+// the public key information from the 'jku' header in order to retrieve key information
+// (key ID, RSA public key), which is used to verify the token.
+//
+// Limitation: only supports RSA tokens using the 'jku' header, which points to a URL
+// that can be used to retrieve key information.
+func (br *BoshResolver) parseRsaToken(token *jwt.Token) (interface{}, error) {
+	if rsa, ok := token.Method.(*jwt.SigningMethodRSA); ok {
+		// with the RSA signing method, the key is a public key / certificate that can be
+		// retrieved from the JKU endpoint (among other places).
+		if rawKeyInfoURL, ok := token.Header["jku"].(string); ok {
+			var kid string
+			if kid, ok = token.Header["kid"].(string); ok {
+				keyInfoURL, err := url.Parse(rawKeyInfoURL)
+				if err != nil {
+					return nil, err
+				}
+
+				key, err := br.fetchPublicKey(keyInfoURL, kid)
+				if err != nil {
+					return nil, err
+				}
+
+				if rsa.Alg() != key.Alg {
+					return nil, fmt.Errorf("signature algorithm %q does not match expected token key information %q", rsa.Alg(), key.Alg)
+				}
+
+				// the RSA public key returned here is used to check the JWT token signature.
+				// It is provided by the URL encoded in the token (in the 'jku' header).
+				// For valid tokens, this URL is verified against the UAA URLs reported by BOSH Director later.
+				return jwt.ParseRSAPublicKeyFromPEM([]byte(key.Value))
+			}
+		}
+
+		return nil, fmt.Errorf("could not find key information URL in token headers: %+v", token.Header)
+	}
+
+	return nil, fmt.Errorf("unsupported signing method: %v", token.Header["alg"])
+}
+
+// fetchPublicKey fetches the token key information from url and returns the key with the Key ID (kid).
+//
+// returns an error if no key can be found with the requested kid or an error arises while communicating with url.
+//
+// Limitation: This will only fetch keys with RSA as signature algorithm.
+func (br *BoshResolver) fetchPublicKey(url *url.URL, kid string) (*UaaKeyInfo, error) {
+
+	res, err := br.client.Do(&http.Request{
+		Method: http.MethodGet,
+		URL:    url,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	defer res.Body.Close()
+
+	keys := struct {
+		Keys []UaaKeyInfo
+	}{}
+
+	data, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.Unmarshal(data, &keys)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, key := range keys.Keys {
+		if key.Kty == "RSA" && key.Kid == kid {
+			matchingKey := key
```

Comment:
Do we really need a new variable?

Reply:
My idea was to have a copy of the one value I need, so the full response can be garbage collected.